hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -48,6 +48,17 @@ std::optional<std::pair<ModuleID, StructID>> typeID(T&& x) {
     return {{id->sub(-2), id->sub(-1)}};
 }
 
+// Helper function to detect whether a given type is a filter.
+bool isFilter(const declaration::Type& type) {
+    const auto unit = type.type().as<type::Struct>();
+
+    // By the point we reach the optimizer Spicy unit properties are
+    // not present anymore, so we detect this with a heuristic.
+    //
+    // TODO(bbannier): consider explicitly expressing this information in the AST.
```

Comment:
We are using a couple of HILTI-internal attributes already, could add one like `&no-optimize` that Spicy would add to filter types?

Reply:
The approach in this patch is to progressively empty out units so that we can remove empty units in a final pass (that part is not yet implemented), so we'd have to prevent removal of unit member functions with `&no-optimize`; adding that for all filter functions seems very manual. I'll instead try to make the optimizer smarter about what unit member functions are used, see my comment below.